web/src/bootstrap_typeahead.js

```diff
@@ -0,0 +1,657 @@
+                    // If the typeahead isn't shown yet, the `lookup` call will open it.
+                    // Here we make a callback to the input field before we open the
+                    // lookahead in case it needs to make UI changes first (e.g. widening
+                    // the search bar).
+                    this.options.openInputFieldOnKeyUp();
+                }
+                this.lookup(false);
+        }
+
+        this.maybeStopAdvance(e);
+
+        e.preventDefault();
+    },
+
+    blur(e) {
+        // Blurs that move focus to elsewhere within the parent element shouldn't
+        // hide the typeahead.
+        if (
+            this.options.parentElement &&
+            $(e.relatedTarget).parents(this.options.parentElement).length > 0
+        ) {
+            return;
+        }
+        setTimeout(() => {
+            if (!this.$container.is(":hover")) {
+                this.hide();
+            } else if (this.shown) {
+                // refocus the input if the user clicked on the typeahead
+                // so that clicking elsewhere registers as a blur and hides
+                // the typeahead.
+                this.$element.trigger("focus");
+            }
+        }, 150);
+    },
+
+    element_click() {
+        // update / hide the typeahead menu if the user clicks anywhere
+        // inside the typing area, to avoid misplaced typeahead insertion.
+        this.lookup(false);
+    },
+
+    click(e) {
+        e.stopPropagation();
+        e.preventDefault();
+        // The original bootstrap code expected `mouseenter` to be called
+        // to set the active element before `select()`.
+        // Since select() selects the element with the active class, if
+        // we trigger a click from JavaScript (or with a mobile touchscreen),
+        // this won't work. So we need to set the active element ourselves,
+        // and the simplest way to do that is to just call the `mouseenter`
+        // handler here.
+        this.mouseenter(e);
+        this.select(e);
+    },
+
+    mouseenter(e) {
+        if (!this.mouse_moved_since_typeahead) {
+            // Prevent the annoying interaction where your mouse happens
+            // to be in the space where typeahead will open.  (This would
+            // result in the mouse taking priority over the keyboard for
+            // what you selected). If the mouse has not been moved since
+            // the appearance of the typeahead menu, we disable the
+            // cursor, which in turn prevents the currently hovered
+            // element from being selected.  The mousemove handler
+            // overrides this logic.
+            $(e.currentTarget).find("a").css("cursor", "none");
+            return;
+        }
+        this.$menu.find(".active").removeClass("active");
+        $(e.currentTarget).addClass("active");
+    },
+};
+
+/* TYPEAHEAD PLUGIN DEFINITION
+ * =========================== */
+
+$.fn.typeahead = function (option) {
```

Comment:
Another followup that will aid the TypeScript conversion would be to change this from a jQuery plugin to a normal exported function. (While it’s possible to declare an annotation that assumes this exists as a jQuery plugin, all unchecked assumptions weaken type safety.)